File: frontend/lib/api/supervision.ts
import apiClient from './client';

export interface SupervisionPerson {
  id: number;
  name: string;
  username: string;
  email?: string;
  phone?: string;
  designation?: string;
  department?: string;
  training_site?: string;
}

export interface SupervisionAssignment {
  id: number;
  assignment_type: 'PRIMARY' | 'CO_SUPERVISOR';
  status: 'ACTIVE' | 'ENDED' | 'SUSPENDED';
  start_date: string | null;
  end_date: string | null;
  notes?: string;
  reason_for_change?: string;
  supervisor: SupervisionPerson;
  resident: SupervisionPerson;
}

export interface SupervisionAssignmentPayload {
  resident_id: number;
  supervisor_id: number;
  assignment_type?: 'PRIMARY' | 'CO_SUPERVISOR';
  start_date: string;
  notes?: string;
}

export interface SupervisionEndPayload {
  end_date: string;
  reason_for_change?: string;
}

export interface SupervisionChangePrimaryPayload {
  resident_id: number;
  new_supervisor_id: number;
  start_date: string;
  reason_for_change?: string;
}

export interface SupervisionOptionPerson {
  id: number;
  name: string;
  username: string;
  training_site?: string;
  department?: string;
  designation?: string;
  has_active_primary?: boolean;
  active_primary_count?: number;
  active_total_count?: number;
}

export interface SupervisionOptions {
  residents: SupervisionOptionPerson[];
  supervisors: SupervisionOptionPerson[];
}

export type SupervisionDataQuality = Record<string, Array<Record<string, unknown>>>;

export interface SupervisionImportResult {
  success: boolean;
  dry_run: boolean;
  successes: Array<Record<string, unknown>>;
  failures: Array<Record<string, unknown>>;
  warnings?: Array<Record<string, unknown>>;
}

export const supervisionApi = {
  listAssignments: async (params?: Record<string, unknown>) => {
    const response = await apiClient.get<{ count?: number; results?: SupervisionAssignment[] } | SupervisionAssignment[]>(
      '/api/supervision/assignments/',
      { params }
    );
    return Array.isArray(response.data) ? response.data : response.data.results || [];
  },
  getAssignment: async (id: number) => {
    const response = await apiClient.get<SupervisionAssignment>(`/api/supervision/assignments/${id}/`);
    return response.data;
  },
  createAssignment: async (payload: SupervisionAssignmentPayload) => {
    const response = await apiClient.post<SupervisionAssignment>('/api/supervision/assignments/', payload);
    return response.data;
  },
  endAssignment: async (id: number, payload: SupervisionEndPayload) => {
    const response = await apiClient.post<SupervisionAssignment>(`/api/supervision/assignments/${id}/end/`, payload);
    return response.data;
  },
  changePrimary: async (payload: SupervisionChangePrimaryPayload) => {
    const response = await apiClient.post<SupervisionAssignment>('/api/supervision/change-primary/', payload);
    return response.data;
  },
  getSupervisionOptions: async (params?: Record<string, unknown>) => {
    const response = await apiClient.get<SupervisionOptions>('/api/supervision/options/', { params });
    return response.data;
  },
  getSupervisionDataQuality: async () => {
    const response = await apiClient.get<SupervisionDataQuality>('/api/supervision/data-quality/');
    return response.data;
  },
  importSupervisionCsv: async (file: File, dryRun = true) => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('dry_run', dryRun ? 'true' : 'false');
    const response = await apiClient.post<SupervisionImportResult>('/api/supervision/import/', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },
};

export default supervisionApi;

File: android/app/src/main/java/fmu/pg/sims/core/model/Supervision.kt
package fmu.pg.sims.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** GET /api/supervision/options/ */
@Serializable
data class SupervisionOptionsResponse(
    @SerialName("supervisors") val supervisors: List<SupervisorOption> = emptyList(),
)

@Serializable
data class SupervisorOption(
    @SerialName("id") val id: Int,
    @SerialName("name") val name: String = "",
    @SerialName("training_site") val trainingSite: String = "",
    @SerialName("department") val department: String = "",
    @SerialName("designation") val designation: String = "",
    @SerialName("active_primary_count") val activePrimaryCount: Int = 0,
    @SerialName("active_total_count") val activeTotalCount: Int = 0,
)

@Serializable
data class ResidentSummaryDto(
    @SerialName("id") val id: Int,
    @SerialName("name") val name: String = "",
    @SerialName("username") val username: String = "",
    @SerialName("department") val department: String = "",
    @SerialName("training_site") val trainingSite: String = "",
)

@Serializable
data class SupervisorSummaryDto(
    @SerialName("id") val id: Int,
    @SerialName("name") val name: String = "",
    @SerialName("department") val department: String = "",
    @SerialName("training_site") val trainingSite: String = "",
    @SerialName("designation") val designation: String = "",
)

/** GET/POST /api/supervision/assignments/ */
@Serializable
data class ResidentSupervisorAssignmentDto(
    @SerialName("id") val id: Int = 0,
    @SerialName("resident") val resident: ResidentSummaryDto? = null,
    @SerialName("supervisor") val supervisor: SupervisorSummaryDto? = null,
    @SerialName("assignment_type") val assignmentType: String = "PRIMARY",
    @SerialName("status") val status: String = "ACTIVE",
    @SerialName("is_active") val isActive: Boolean = true,
    @SerialName("start_date") val startDate: String? = null,
    @SerialName("end_date") val endDate: String? = null,
    @SerialName("notes") val notes: String = "",
)

/** POST body for /api/supervision/assignments/ (create). */
@Serializable
data class CreateSupervisorAssignmentRequest(
    @SerialName("resident_id") val residentId: Int,
    @SerialName("supervisor_id") val supervisorId: Int,
    @SerialName("assignment_type") val assignmentType: String = "PRIMARY",
    @SerialName("start_date") val startDate: String,
    @SerialName("notes") val notes: String = "",
)

/** POST /api/pending-supervisor-links/ — "supervisor not listed" request. Never fabricates a
 * SupervisorProfile; resolved later by an administrator. */
@Serializable
data class CreatePendingSupervisorLinkRequest(
    @SerialName("resident") val resident: Int,
    @SerialName("supervisor_name_text") val supervisorNameText: String,
    @SerialName("department_text") val departmentText: String = "",
    @SerialName("institution_text") val institutionText: String = "",
    @SerialName("pmdc_number_text") val pmdcNumberText: String = "",
    @SerialName("email_text") val emailText: String = "",
    @SerialName("phone_text") val phoneText: String = "",
    @SerialName("notes") val notes: String = "",
)

@Serializable
data class PendingSupervisorLinkDto(
    @SerialName("id") val id: Int,
    @SerialName("resident") val resident: Int,
    @SerialName("supervisor_name_text") val supervisorNameText: String = "",
    @SerialName("status") val status: String = "PENDING",
)
